sys/mips/82371AB.c

```diff
@@ -0,0 +1,141 @@
+/* 82371AB PCI-ISA bridge driver */
+
+#define KL_LOG KL_DEV
+#include <sys/mimiker.h>
+#include <sys/device.h>
+#include <sys/rman.h>
+#include <sys/bus.h>
+#include <mips/malta.h>
+#include <sys/pci.h>
+#include <sys/interrupt.h>
+#include <dev/piixreg.h>
+#include <dev/isareg.h>
+#include <sys/libkern.h>
+#include <sys/devclass.h>
+
+typedef struct intel_isa_state {
+  resource_t *io;
+  rman_t io_rman;
+} intel_isa_state_t;
+
+static resource_t *intel_isa_alloc_resource(device_t *dev, res_type_t type,
+                                            int rid, rman_addr_t start,
+                                            rman_addr_t end, size_t size,
+                                            rman_flags_t flags) {
+  assert(dev->bus == DEV_BUS_ISA);
+
+  if (type != RT_IOPORTS) {
+    device_t *idev = BUS_METHOD_PROVIDER(dev->parent, alloc_resource);
+    return BUS_METHODS(idev->parent)
+      .alloc_resource(idev, type, rid, start, end, size, flags);
+  }
+
+  assert(start >= IO_ICUSIZE);
+
+  rman_t *rman = NULL;
+  intel_isa_state_t *isa = dev->parent->state;
+  rman = &isa->io_rman;
+  bus_space_handle_t bh = isa->io->r_bus_handle;
+
+  resource_t *r =
+    rman_reserve_resource(rman, type, rid, start, end, size, 0, flags);
+
+  if (r == NULL)
+    return NULL;
+
+  /* We shouldn't do that for IRQs, but rn we'll never end up here when
+   * allocating them, because they get dispatched. */
+  r->r_bus_tag = generic_bus_space;
+  r->r_bus_handle = bh + resource_start(r);
+
+  /* IOPorts are always active */
+  if (bus_activate_resource(dev, r)) {
+    resource_release(r);
+    return NULL;
+  }
+
+  return r;
+}
+
+static int intel_isa_activate_resource(device_t *dev, resource_t *r) {
+  /* Neither IRQs in general or IOports on ISA require activation */
+  return 0;
+}
+
+static void intel_isa_deactivate_resource(device_t *dev, resource_t *r) {
+  /* Neither IRQs in general or IOports on ISA require deactivation */
```

Comment:
The same problem as with the activation.

Reply:
Yes, and my answer is the same too.